eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,476 @@
+			for {
+				select {
+				case <-ctx.Done():
+					return nil, ctx.Err()
+				case <-ec.closed:
+					return nil, fmt.Errorf("client closed")
+				case err, ok := <-singleErrCh:
+					if !ok {
+						// If the error channel closed, treat that as no more logs or success.
+						break processErrsLabel
+					}
+					fromBlock = max(fromBlock, lastBlock+1)
+					return nil, err // Try another client
+				}
+			}
+
+			return nil, nil
+		}
+
+		if _, err := ec.call(contextWithMethod(ctx, "FetchHistoricalLogs [process logs]"), processLogsFunc); err != nil {
+			errCh <- err
+		}
+	}()
+
+	return logsCh, errCh, nil
+}
+
+// StreamLogs subscribes to events emitted by the contract.
+func (ec *MultiClient) StreamLogs(ctx context.Context, fromBlock uint64) <-chan BlockLogs {
+	logs := make(chan BlockLogs)
+
+	go func() {
+		defer close(logs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ec.closed:
+				return
+			default:
+				// Update healthiness of all nodes and make sure at least one of them is available.
+				if err := ec.Healthy(ctx); err != nil {
+					ec.logger.Fatal("no healthy clients", zap.Error(err))
+				}
+
+				f := func(client SingleClientProvider) (any, error) {
+					lastBlock, err := client.streamLogsToChan(ctx, logs, fromBlock)
+					if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+						return lastBlock, err
+					}
+					if err != nil {
+						// fromBlock's value in the outer scope is updated here, so this function needs to be a closure
+						fromBlock = max(fromBlock, lastBlock+1)
+						return nil, err
+					}
+
+					return nil, nil
+				}
+
+				_, err := ec.call(contextWithMethod(ctx, "StreamLogs"), f)
+				if err != nil && !errors.Is(err, ErrClosed) && !errors.Is(err, context.Canceled) {
+					// NOTE: There are unit tests that trigger Fatal and override its behavior.
+					// Therefore, the code must call `return` afterward.
+					ec.logger.Fatal("failed to stream registry events", zap.Error(err))
+				}
+				return
+			}
+		}
+	}()
+
+	return logs
+}
+
+// Healthy returns if execution client is currently healthy: responds to requests and not in the syncing state.
+func (ec *MultiClient) Healthy(ctx context.Context) error {
+	var atLeastOneHealthy atomic.Bool
+	var errs error
+	var errsMu sync.Mutex
+
+	var wg sync.WaitGroup
+	// We need to fetch all clients to update the internal state
+	for i, client := range ec.clients {
+		wg.Add(1)
+		go func(i int, client SingleClientProvider) {
+			defer wg.Done()
+
+			if err := client.Healthy(ctx); err != nil {
+				ec.logger.Warn("client is not healthy",
+					zap.String("addr", ec.nodeAddrs[i]),
+					zap.Error(err))
+
+				errsMu.Lock()
+				errs = errors.Join(errs, err)
+				errsMu.Unlock()
+			} else {
+				ec.logger.Debug("client is healthy",
+					zap.String("addr", ec.nodeAddrs[i]))
+
+				atLeastOneHealthy.Store(true)
+			}
+		}(i, client)
+	}
+	wg.Wait()
+
+	if !atLeastOneHealthy.Load() {
+		return errs
+	}
+
+	return nil
+}
```

Comment:
isn't it the same as just using `call` as before which should also return no error only if at least one is healthy?

Reply:
@moshe-blox if all clients are healthy, `call` would check only one client. The current `Healthy` implementation calls `Healthy` on all clients to update their state

Reply:
i see, that makes sense, thanks 👍

can you pls try using https://pkg.go.dev/github.com/sourcegraph/conc/pool#ErrorPool for this? i imagine it can simplify the code here quite a bit

it should handle concurrency and join the errors for you

plus you can just check that the err returned by `Wait` isn't nil instead of having `atLeastOneHealth`

Reply:
@moshe-blox great idea, as I understand it's similar to https://pkg.go.dev/golang.org/x/sync/errgroup, right? Do you have any objections against using `errgroup` instead?

Reply:
Actually, I'm not sure if it will look better because we need to return no error if at least one client is healthy, so it would require parsing the combined error